### PR TITLE
Genode build integration

### PR DESCRIPTION
**Pull request by ehmry**
These patches are for building the Genode bindings and not simply a stub library. Building the final product here is useful because it would allow me to automatically test Solo5 PRs.

I've disabled `test_tls` for lack of TLS. I've also added the ability to override `HOSTCC` via an environmental variable when running `configure.sh`, this is to allow the bindings and elftool to be built for different platforms (the former the target, the latter the build machine).

### CODE REVIEW

**Comment by ehmry** [comment]
As I mentioned there is some CI now in place, and these patches would allow me to test future PRs without applying patches: https://hydra.hq.c3d2.de/project/genodepkgs (IPv6 only)
hydra

File: include/solo5/mft_abi.h [hunk 1]
```diff
@@ -154,7 +154,8 @@ struct mft1_note {
  * Internal alignment of (m) within struct mft1_note. Must be passed to
  * elf_load_note() as note_align when loading.
  */
-#define MFT1_NOTE_ALIGN offsetof(struct { char c; struct mft m; }, m)
+struct _mft1_note_aligned { char c; struct mft m; };
+#define MFT1_NOTE_ALIGN offsetof(struct _mft1_note_aligned, m)
```
[the remark this answers is not in the document]
Comment (ehmry) [reply include/solo5/mft_abi.h]: I assume its a matter of how Clang enforces the C++ spec. Fortunately this seems to be only awkward part of including the headers as C++.

File: Makefile.common [hunk 2]
```diff
@@ -32,6 +32,7 @@ include $(TOPDIR)/Makeconf
 CC := $(MAKECONF_CC)
 CFLAGS := -std=c11 -Wall -Wextra -Werror -O2 -g
 CFLAGS += -ffreestanding -fstack-protector-strong $(MAKECONF_CFLAGS)
+CXXLAGS += -Wall -Wextra -Werror -O2 -g
```
[the remark this answers is not in the document]
Comment (ehmry) [reply Makefile.common]: Good catch.

File: Makefile.common [hunk 3]
```diff
@@ -53,6 +54,12 @@ define COMPILE.c
 	mv -f $*.Td $*.d && touch $@
 endef
 
+define COMPILE.cc
+	@echo "CXX $<"
+	$(CXX) $(DEPFLAGS) $(CXXFLAGS) $(CPPFLAGS) -c $< -o $@
```
[the remark this answers is not in the document]
Comment (ehmry) [reply Makefile.common]: That can be added to `configure.sh` without problems for other platforms.

File: GNUmakefile [hunk 4]
```diff
@@ -100,7 +100,8 @@ install-opam-%: all opam/solo5-bindings-%.pc force-install
 	    $(PREFIX)/lib/solo5-bindings-$* \
 	    $(PREFIX)/include/solo5-bindings-$*/solo5 \
 	    $(PREFIX)/include/solo5-bindings-$*/crt
-	cp -R include/solo5 include/crt $(PREFIX)/include/solo5-bindings-$*
+	cp -R include/solo5 $(PREFIX)/include/solo5-bindings-$*
+	if [ -d "include/crt" ]; then cp -R include/crt $(PREFIX)/include/solo5-bindings-$*; fi
```
[the remark this answers is not in the document]
Comment (ehmry) [reply GNUmakefile]: Yes, thats pretty clean to do.

File: include/solo5/mft_abi.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (mato) [reply include/solo5/mft_abi.h]: Ah right, you're including those using a C++ compiler. Got it. The construct should be equivalent, so I think this is fine.

**Comment by mato** [comment]
@ehmry I don't follow how this change is intended to work in the case where someone wants to build Solo5 (MirageOS) unikernels _for_ Genode on a host _without_ the Genode toolchain. You've now disabled `CONFIG_GENODE` for those cases, so nothing will be built.

In practice, that seems to mean that there's now no workable path to `mirage configure -t genode` via OPAM, or am I misunderstanding something?

**Comment by ehmry** [comment]
I'm not sure, until I can answer that question this issue is blocked.

**Comment by mato** [comment]
As mentioned in https://github.com/Solo5/solo5/pull/446#issuecomment-612595136, there seems to be no way forward for this right now, so closing.

**Pull request closed without merging**